docs/design/README.md

```diff
@@ -0,0 +1,130 @@
+
+kraan controller is the main controller that is responsible to watching AddonLayer
+custom resources and bringing to its desired state. In attempting to do so, it 
+relies on helm-operator to deploy the addon helm charts that are part of the layer. Helm operator watches
+and reconciles *HelmRelease* custom resource. Each *HelmRelease* custom resources
+represents a single addon. For more details on *HelmRelease*, check 
+[here](https://docs.fluxcd.io/projects/helm-operator/en/stable/helmrelease-guide/introduction/).
+source-controller is one of the components in [gitops-toolkit](https://toolkit.fluxcd.io/) which 
+helps abstracting away git interaction from kraan. By design, an AddonLayer will
+point to a *directory in a git repository* which contains the list of
+HelmReleases. Whenever kraans' reconciliation logic is kicked off, it order to
+fetch the list of helm releases that are part of that addon layer, it will reach out
+to source-controller api to fetch repo files instead of reaching out to git directly. 
+
+
+***Note:*** *In future, kraan will support packaging addons in formats other 
+than helm charts e.g kustomization*
+ 
+### Spec
+The addons are first packaged as HelmRelease custom resources which represents
+a helm chart. A sample helm release custom resource is shown below.
+For more details on *HelmRelease*, check
+[here](https://docs.fluxcd.io/projects/helm-operator/en/stable/helmrelease-guide/introduction/).
+
+
+```yaml
+# podinfo.yaml
+---
+apiVersion: helm.fluxcd.io/v1
+kind: HelmRelease
+metadata:
+  name: podinfo
+  namespace: default
+spec:
+  chart:
+    repository: https://stefanprodan.github.io/podinfo
+    name: podinfo
+    version: 3.2.0
+```
+
+###### Example
+A sample addon layer custom resource is shown below. 
+```yaml
+apiVersion: kraan.io/v1alpha1
+kind: AddonsLayer
+metadata:
+  name: bootstrap
+
+spec:
+  version: 1.0.0
+  hold: true
+  interval: 1m
+  source:
+    name: addons
+    namespace: gitops-system
+    path: ./addons/bootstrap
+  prereqs:
+      k8sVersion: "1.16"
+---
+apiVersion: kraan.io/v1alpha1
+kind: AddonsLayer
+metadata:
+  name: base
+
+spec:
+  version: 1.0.1
+  interval: 1m
+  source: 
+    name: addons
+    namespace: gitops-system
+    path: ./addons/base
+  prereqs:
+      k8sVersion: "1.16"
+      dependsOn:
+        - bootstrap@1.0.0
+```
+
+## Reconciliation Logic
+Kraan controller is built using the [kubebuilder](https://github.com/kubernetes-sigs/kubebuilder) project.
+From a controllers' perspective, reconciliation is where it attempts to bring the
+current state of the layer into its desired state. The reconciliation logic is 
+kicked off whenever there is a change detected in the AddonLayer custom resource.
+It could be an addition, deletion or an update to the resource.Following is 
+the step by step reconciliation logic in kraan. Whenever an event is 
+received for an AddonLayer,
+
+* check whether prereqs.k8sVersion == cluster version. If the cluster is not
```

Comment:
Check whether the version of the Kubernetes API deployed on the cluster is greater than or equalt to prereqs.K8sVersion